widget/embedded/src/lib/tabManager/tabManager.ts

```diff
@@ -0,0 +1,223 @@
+import type {
+  Events,
+  ID,
+  Message,
+  TabManagerInterface,
+} from './tabManager.types';
+
+import { CHANNEL_NAME, MAXIMUM_ID, TIMEOUT } from './tabManager.constants';
+
+export class TabManager implements TabManagerInterface {
+  private tabId: ID;
+  private state: 'not-initiated' | 'not-claimed' | 'claimed' = 'not-initiated';
+  private lastTryClaim: number | undefined;
+  private events: Events = {};
+  private channel: BroadcastChannel;
+
+  constructor(events: Events) {
+    this.channel = new BroadcastChannel(CHANNEL_NAME);
+    this.tabId = Math.trunc(Math.random() * MAXIMUM_ID);
+    this.events = events;
+    this.initEvents();
+    void this.tryClaim();
+  }
+
+  /**
+   * This is applicable when a user in an inactive tab seeks to activate that tab.
+   */
+  forceClaim = () => {
+    if (!this.isClaimed()) {
+      const message: Message = { name: 'force-claim', candidateId: this.tabId };
+      this.channel.postMessage(message);
+      setTimeout(() => {
+        if (!this.isClaimed()) {
+          this.claim();
+        }
+      }, TIMEOUT);
+    }
+  };
+
+  isClaimed = () => {
+    return this.state === 'claimed';
+  };
+
+  /**
+   * remove event listeners
+   */
+  destroy = () => {
+    this.channel.removeEventListener('message', this.handleMessageEvent);
+    document.removeEventListener(
+      'visibilitychange',
+      this.handleVisibilityChange
+    );
+    document.removeEventListener('resume', this.handleResume);
+  };
+
+  /**
+   * Attach some event listeners to the window.
+   * It includes listening to broadcast messages, visibilitychange and resume event.
```

Comment:
The implementation can change in future, not including the exact event name would be better.

Something like this.
```
Reacting to messages and window events to claim.
```